### PR TITLE
feat(abstract-utxo): add psbt and musig2 support for sdk-api

### PR DESCRIPTION
**Pull request by saravanan7mani**
Adding PSBT and taproot key path spend support for the `prebuildAndSign` method.

Added a unit test for the `prebuildAndSign` method as well as unit tests in `utxo-lib` for the utility methods that we implemented there

BG-66947

### CODE REVIEW

File: modules/abstract-utxo/src/abstractUtxoCoin.ts [hunk 1]
```diff
@@ -720,6 +735,13 @@ export abstract class AbstractUtxoCoin extends BaseCoin {
     params: VerifyTransactionOptions<TNumber>
   ): Promise<boolean> {
     const { txParams, txPrebuild, wallet, verification = { allowPaygoOutput: true }, reqId } = params;
+
+    // TODO: BG-77743 Add native psbt support
+    if (txPrebuild.txHex && bitgo.isPsbt(txPrebuild.txHex)) {
+      const psbt = bitgo.createPsbtFromHex(txPrebuild.txHex, this.network);
+      txPrebuild.txHex = psbt.getUnsignedTx().toHex();
+    }
```
Comment (OttoAllmendinger) [inline modules/abstract-utxo/src/abstractUtxoCoin.ts:743]: can we do this inside `parseTransaction` instead?

File: modules/bitgo/test/v2/unit/coins/utxo/transaction.ts [hunk 2]
```diff
@@ -44,10 +43,30 @@ function run<TNumber extends number | bigint = number>(
   amountType: 'number' | 'bigint' = 'number'
 ) {
   describe(`Transaction Stages ${coin.getChain()} (${amountType}) scripts=${inputScripts.join(',')}`, function () {
-    const wallet = getUtxoWallet(coin);
-    const walletKeys = getDefaultWalletKeys();
+    const isTransactionWithKeyPathSpend = inputScripts.some((s) => s === 'p2trMusig2');
+    const isPsbt = isTransactionWithKeyPathSpend;
+    const isTransactionWithReplayProtection = inputScripts.some((s) => s === 'replayProtection');
```
Comment (OttoAllmendinger) [inline modules/bitgo/test/v2/unit/coins/utxo/transaction.ts:48]: Let's make `isPsbt` a new parameter here instead. We want to cover the other input scripts as well.

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 3]
```diff
@@ -511,18 +511,39 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
+      : input.bip32Derivation?.length
+      ? UtxoPsbt.deriveKeyPair(bip32Interface, input.bip32Derivation)?.publicKey
+      : undefined;
+    if (!pubKey) {
+      throw new Error('can not derive from HD key pair');
+    }
+    return this.validateSignaturesOfInputCommon(inputIndex, pubKey);
+  }
+
+  validateSignaturesOfInputCommon(inputIndex: number, pubkey?: Buffer): boolean {
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts:532]: `*Inner` was a private method. Now it is made as public, so `*Inner` is misleading.

File: modules/bitgo/test/v2/unit/coins/utxo/transaction.ts [hunk 4]
```diff
@@ -137,9 +187,17 @@ function run<TNumber extends number | bigint = number>(
     });
 
     it('match fixtures', async function () {
+      if (isPsbt) {
+        // TODO (maybe) - once PSBT is enabled for all script types, custom JSON representation of PSBT can be created and tested here.
+        // signatures of taprootKeyPathSpends are not random since random nature of MuSig2 nonce, so psbt hex comparison also wont work.
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/bitgo/test/v2/unit/coins/utxo/transaction.ts]: Done

File: modules/bitgo/test/v2/unit/coins/utxo/transaction.ts (hunk 2 above)
Comment (saravanan7mani) [reply modules/bitgo/test/v2/unit/coins/utxo/transaction.ts]: Done

File: modules/abstract-utxo/src/abstractUtxoCoin.ts (hunk 1 above)
Comment (saravanan7mani) [reply modules/abstract-utxo/src/abstractUtxoCoin.ts:743]: `verifyTransanction` is also called from wallet.ts (common to coins) too in our flow where we don't want this.

File: modules/sdk-core/src/bitgo/wallet/iWallet.ts [hunk 5]
```diff
@@ -114,6 +114,7 @@ export interface PrebuildTransactionOptions {
    * If set to false, sweep all funds including the required minimums for address(es). E.g. Polkadot (DOT) requires 1 DOT minimum.
    */
   keepAlive?: boolean;
+  txFormat?: string;
```
Comment (pengyuc-bitgo) [inline modules/sdk-core/src/bitgo/wallet/iWallet.ts:117]: This name is quite generic but it seems its just for UTXO now. can you put some comment on this?

Comment (pengyuc-bitgo) [reply modules/sdk-core/src/bitgo/wallet/iWallet.ts:117]: Should we put a type on this field?

Comment (saravanan7mani) [reply modules/sdk-core/src/bitgo/wallet/iWallet.ts]: Done

**Review by pengyuc-bitgo — DISMISSED** [review]

**Review by angshal — DISMISSED** [review]

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/abstract-utxo/src/abstractUtxoCoin.ts [hunk 6]
```diff
@@ -719,7 +731,23 @@ export abstract class AbstractUtxoCoin extends BaseCoin {
   async verifyTransaction<TNumber extends number | bigint = number>(
     params: VerifyTransactionOptions<TNumber>
   ): Promise<boolean> {
-    const { txParams, txPrebuild, wallet, verification = { allowPaygoOutput: true }, reqId } = params;
+    const { txParams, txPrebuild: txPrebuildObj, wallet, verification = { allowPaygoOutput: true }, reqId } = params;
+    let txPrebuild;
+
+    // TODO: BG-77743 Add native psbt support - we make a deep copy of the txPrebuild object here to avoid
+    // modifying the original object so that we can still use it later on in the function
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/abstract-utxo/src/abstractUtxoCoin.ts:738]: Done

File: modules/bitgo/test/v2/unit/coins/utxo/transaction.ts [hunk 7]
```diff
@@ -136,10 +200,20 @@ function run<TNumber extends number | bigint = number>(
       transactionStages = await getTransactionStages();
     });
 
+    afterEach(nock.cleanAll);
+
     it('match fixtures', async function () {
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/bitgo/test/v2/unit/coins/utxo/transaction.ts]: Done

File: modules/sdk-core/src/bitgo/wallet/iWallet.ts [hunk 8]
```diff
@@ -114,6 +114,11 @@ export interface PrebuildTransactionOptions {
    * If set to false, sweep all funds including the required minimums for address(es). E.g. Polkadot (DOT) requires 1 DOT minimum.
    */
   keepAlive?: boolean;
+  /**
+   * This comment applies to UTXO coins. It's asking which transaction format to use:
+   * the standard network format (also known as 'legacy' or 'default'), or the 'psbt' format, which follows the BIP-174.
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/sdk-core/src/bitgo/wallet/iWallet.ts]: Done

File: modules/abstract-utxo/src/abstractUtxoCoin.ts [hunk 9]
```diff
@@ -68,7 +68,7 @@ const debug = debugLib('bitgo:v2:utxo');
 
 import ScriptType2Of3 = utxolib.bitgo.outputScripts.ScriptType2Of3;
 import { isReplayProtectionUnspent } from './replayProtection';
-import { signAndVerifyWalletTransaction } from './sign';
+import { signAndVerifyPsbt, signAndVerifyWalletTransaction } from './sign';
```
Comment (OttoAllmendinger) [inline modules/abstract-utxo/src/abstractUtxoCoin.ts:71]:
```suggestion
import { signAndVerifyPsbt, signAndVerifyLegacyTransaction } from './sign';
```

**Review by OttoAllmendinger — APPROVED** [review]
rename can happen in follow up

**Pull request merged**